### PR TITLE
document support for user-defined hooks (WIP)

### PR DESCRIPTION
**Pull request by boegel**
documentation update for feature implemented in https://github.com/easybuilders/easybuild-framework/pull/2343

preview at http://boegel-eb.readthedocs.io/en/hooks_docs/Hooks.html

### CODE REVIEW

File: docs/Hooks.rst [hunk 1]
```diff
@@ -0,0 +1,143 @@
+.. _hooks:
+
+Hooks
+=====
+
+Since EasyBuild v3.5.0, support is available for implementing *hooks* that can be used
+to customise the behaviour of EasyBuild according to site policies if needed,
+without having having to change the EasyBuild framework or the existing easyblocks.
```
Comment (damianam) [inline docs/Hooks.rst]: having having?

File: docs/Hooks.rst [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+*Hooks* are user-defined functions that are called by the EasyBuild framework at specific times during
+the installation procedure that is performed for each software package being installed.
+
+They can be leveraged to alter or augment the installation procedure in a particular way.
+This is usually done to conform with site-specific policies that are difficult to enforce otherwise,
+but it can also be (ab)used to fix specific problems or inject self-implemented enhancements
+(before you flesh them out in a proper contribution, for example).
+
+Both the ``EasyBlock`` instance and the parsed easyconfig file that are being used
+are fully accessible (and modifiable) from within hook implementations, Hence, this mechanism
```
Comment (damianam) [inline docs/Hooks.rst]: `implementations. Hence`

File: docs/Hooks.rst [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+
+.. _hooks_available:
+
+Available hooks
+---------------
+
+Currently, two types of hooks are supported:
+
+* ``start_hook`` and ``end_hook``, which are triggered before starting a new software installation, and right after,
```
Comment (damianam) [inline docs/Hooks.rst]: I thought those hooks were called at the beginning and end of the installation "session" so to say. Not per package. I guess I was wrong, but probably that makes sense. In our case, for instance, we would like to refresh the cache at the end of a "session", not every time a new dependency is installed.

File: docs/Hooks.rst [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  respectfully;
+* hooks that a triggered before and after every step of the installation procedure, aptly named
```
Comment (damianam) [inline docs/Hooks.rst]: hooks that are triggered

File: docs/Hooks.rst (hunk 3 above)
Comment (boegel) [reply docs/Hooks.rst]: I can think of use cases for both, so we should add support for an additional set of hooks `start_session_hook` and `end_session_hook`?

That may call for renaming the existing `start_hook` and `end_hook` to something less ambiguous, like `start_easyconfig_hook` and `end_easyconfig_hook`?

I'm open to suggestions for better names...

Comment (damianam) [reply docs/Hooks.rst]: `init_hook` and `finalize_hook`? And `pre_easyconfig_hook`/`post_easyconfig_hook`?

File: docs/Hooks.rst [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+.. _hooks_what:
+
+What are hooks?
+---------------
+
+*Hooks* are user-defined functions that are called by the EasyBuild framework at specific times during
+the installation procedure that is performed for each software package being installed.
+
+They can be leveraged to alter or augment the installation procedure in a particular way.
```
Comment (damianam) [inline docs/Hooks.rst]: in any way?

**Comment by damianam** [comment]
Let's merge this, lgtm

**Pull request merged**